test/functorch/test_control_flow.py

```diff
@@ -1136,6 +1137,262 @@ def false_fn(x):
             expected_grads = torch.autograd.grad(fn(x), (x,), grad_out)
             self.assertEqual(expected_grads, grads)
 
+    def _test_cond_autograd(self, cond_fct, pred, true_fn, false_fn, operands):
+        from torch.fx.passes.shape_prop import _extract_tensor_metadata, TensorMetadata
+
+        # This is a helper function that extracts the metadata from the tensor and
+        # sets the requries_grad flag to false. This is needed as we compare the
+        # metadata of the operands and the gradients
+        def _extract_tensor_metadata_except_requires_grad(arg):
+            metadata = _extract_tensor_metadata(arg)
+            metadata = TensorMetadata(
+                metadata.shape,
+                metadata.dtype,
+                False,
+                metadata.stride,
+                metadata.memory_format,
+                metadata.is_quantized,
+                metadata.qparams,
+            )
+            return metadata
+
+        # Comparison of FWD path
+        y_cond = cond_fct(pred(*operands), true_fn, false_fn, operands)
+        operands_forced_grad = [o.clone().detach() for o in operands]
+        for o in operands_forced_grad:
+            o.requires_grad = True
+        y_exp = true_fn(*operands_forced_grad) if pred(*operands_forced_grad) else false_fn(*operands_forced_grad)
+        self.assertEqual(y_cond, y_exp)
+
+        # Comparison of BWD path
+        grad_operands = [o for o in operands if o.requires_grad]
+        grad_exp_operands = [o for o in operands_forced_grad if o.requires_grad]
+        
+        # Check if at least some operators require grads
+        if len(grad_operands) > 0:
+            grad_cond = torch.autograd.grad(y_cond, grad_operands, allow_unused=True, retain_graph=True)
+            grad_exp = torch.autograd.grad(y_exp, grad_exp_operands, allow_unused=True, materialize_grads=True)
```

Comment:
why we need additional flags like `retain_graph` and `materialize_grads` and the flags are also different in these two autograd.grad calls?

Reply:
The `retain_graph` flag we need in order to check the graph later one outside of `_test_cond_autograd`. This is only needed for the torch.cond and not for the reference implementation.
The `materialize_grads` flag is needed for the expected gradients. This also returns all-zeros instead of `None`. This flag is only needed for the reference implementation and not for the torch.cond